### PR TITLE
feat(#574): reimplementation of `Image` widget `layout` function

### LINKED ISSUE #574: Masonry: Reason about `Image` widget layout implementation
This is the issue from the TODO comment at:
https://github.com/linebender/xilem/blob/7be4dd2a2f8bf820fc19a3af2a00600d05e4d4c7/masonry/src/widget/image.rs#L94-L97

In particular, the current layout logic entirely ignores the [`FillStrat`](https://github.com/linebender/xilem/blob/7be4dd2a2f8bf820fc19a3af2a00600d05e4d4c7/masonry/src/widget/mod.rs#L62). Effectively, the [`aspect-ratio`](https://developer.mozilla.org/en-US/docs/Web/CSS/aspect-ratio) is always `auto`, so the only troublesome case is if we have a minimum/maximum set which never matches the given aspect ratio.

This logic has recently (#571) been changed from the original [Druid version](https://github.com/linebender/druid/blob/4fe393927f8e59baacb53527a665980a157a0ad4/druid/src/widget/image.rs#L194-L202). 

The steps for this issue are:
1) Work out the correct layout logic. It might be that the current version is correct, at least whilst we can't set `aspect-ratio`.
2) Write a few test cases for the layout logic, and integrate them into the tests module. These could probably use screenshot tests.
3) Remove the original TODO comment.

### PR DESCRIPTION
**Pull request by failingprovince**
Reimplement calculation of image size in `Image` widget `layout` function.
Makes it match against `self.fill` before setting size.

Tests still missing.

Fixes #574

### CODE REVIEW

**Comment by DJMcNab** [comment]
I've edited your comment to include the text "fixes #574", as the version in the title doesn't count.

**Comment by failingprovince** [comment]
Ready for review.
Test looks good to me, but still I'm not sure if they are written correctly.

**Comment by failingprovince** [comment]
There is still one point open: should `FillStrat` be renamed?
If yes, to what?

**Review by jaredoconnell — APPROVED** [review]
I can confirm that this PR does not introduce a regression in the default `Contain` case in the examples. And the tests look good.
Regarding renaming `FillStrat`, I'll defer to the maintainers. I don't love the current name, but it works. So I'm open to either renaming or keeping it.

**Comment by jaredoconnell** [comment]
One comment I have is that there is code in the implementation of `FillStrat` that is not used with this code. What do we want to do about that?

**Comment by failingprovince** [comment]
I think the code of `FillStrat::affine_to_fill` isn't of any use anymore.

Positioning of an image is a work for the `Image::layout` function, plus the current implementation is much simplier than calculating an affine matrix for then defining positioning/layout inside a `BoxConstraint`.

But maybe I'm missing something.

**Comment by DJMcNab** [comment]
Next steps here are removing any unused code, if there is some. I think renaming `FillStrat` to `ObjectFit` is uncontroversial, and I'd recommend doing so in this PR.  We can always rename it again if needed.

**Comment by waywardmonkeys** [comment]
If you rename it, we can remove it from the typos config too!

**Comment by failingprovince** [comment]
> Next steps here are removing any unused code, if there is some. I think renaming `FillStrat` to `ObjectFit` is uncontroversial, and I'd recommend doing so in this PR. We can always rename it again if needed.

Renamed the struct but couldn't find any unused code to remove.

> If you rename it, we can remove it from the typos config too!

Done.

File: masonry/src/widget/image.rs [hunk 1]
```diff
@@ -28,7 +28,7 @@ use crate::{
 /// than the image size).
 pub struct Image {
     image_data: ImageBuf,
-    fill: FillStrat,
+    fill: ObjectFit,
```
Comment (DJMcNab) [inline masonry/src/widget/image.rs:31]: Should we rename this field? I'd suggest either `fit` or `object_fit`?

File: masonry/src/widget/mod.rs [hunk 2]
```diff
@@ -61,7 +61,7 @@ use crate::{Affine, Size};
 // These are based on https://api.flutter.dev/flutter/painting/BoxFit-class.html
 /// Strategies for inscribing a rectangle inside another rectangle.
```
Comment (DJMcNab) [inline masonry/src/widget/mod.rs:62]: Can we link to https://developer.mozilla.org/en-US/docs/Web/CSS/object-fit in these docs?

File: masonry/src/widget/image.rs (hunk 1 above)
Comment (failingprovince) [reply masonry/src/widget/image.rs:31]: Yes we should rename it. 
`object_fit` is ok for the moment, anyway we can always rename it in the future

File: xilem/src/view/image.rs [hunk 3]
```diff
@@ -24,7 +24,7 @@ pub fn image(image: &vello::peniko::Image) -> Image {
         // We take by reference as we expect all users of this API will need to clone, and it's
         // easier than documenting that cloning is cheap.
         image: image.clone(),
-        fill: FillStrat::default(),
+        fill: ObjectFit::default(),
```
Comment (DJMcNab) [inline xilem/src/view/image.rs:27]: Same thing applies here

File: masonry/src/widget/image.rs [hunk 4]
```diff
@@ -28,26 +28,26 @@ use crate::{
...
-            fill: FillStrat::default(),
+            object_fit: ObjectFit::default(),
         }
     }
 
     /// Builder-style method for specifying the fill strategy.
     #[inline]
-    pub fn fill_mode(mut self, mode: FillStrat) -> Self {
-        self.fill = mode;
+    pub fn fill_mode(mut self, mode: ObjectFit) -> Self {
```
Comment (DJMcNab) [inline masonry/src/widget/image.rs:49]: And here

File: masonry/src/widget/image.rs [hunk 5]
```diff
@@ -56,8 +56,8 @@ impl Image {
 impl<'a> WidgetMut<'a, Image> {
     /// Modify the widget's fill strategy.
     #[inline]
-    pub fn set_fill_mode(&mut self, newfil: FillStrat) {
-        self.widget.fill = newfil;
+    pub fn set_fill_mode(&mut self, newfil: ObjectFit) {
```
Comment (DJMcNab) [inline masonry/src/widget/image.rs:59]: And here

**Comment by DJMcNab** [comment]
Thanks!

**Pull request merged**